enhancements/node-tuning/mixed-cpu-node-plugin.md

```diff
@@ -0,0 +1,351 @@
+or `mixedCpus: true` from `workloadHints` reverts the changes from Kubelet `reservedSystemCpus`,
+reboots the nodes that are associated with the updated PerformanceProfile,
+and deletes the mixed-cpu-node-plugin from the cluster.
+
+The feature can be activated/deactivated on a running system.
+
+### Workflow Description
+
+The premise is that OCP cluster is running, and there's an active performance-profile that already tuned the system.
+
+1. The cluster administrator wants to support both exclusive and shared cpus for workloads running on the cluster.
+2. The cluster administrator specifies the shared CPU ids in the PerformanceProfile `spec.cpu.shared`.
+3. The cluster administrator waits for MCO to kick in, update and reboot the nodes.
+4. The cluster administrator waits for the node to come back from reboot.
+5. The application administrator wants to deploy their DPDK application as a Guaranteed pod with shared CPUs.
+6. The application administrator specifies a request for `openshift.io/shared-cpus: 1` under the pod's `spec.containers[].resources.requests`.
+7. The application administrator is waiting for the DPDK pod to be `Running`.
+8. The DPDK's app user/developer wants to run a light-weight task (threads) on shared cpus.
+9. The DPDK's app user/developer should pin the light-weight threads to CPUs that have shown in the `OPENSHIFT_SHARED_CPUS` environment variable of the container's process.
+
+### API Extensions
+N/A
+
+### Risks and Mitigations
+
+* With this solution, the platform's housekeeping processes also runs on the shared CPUs.
+but the intent is to allocate those cpus to workload’s light-weight tasks,
+so having some latency is bearable.
+A way of mitigating that is checks whether it's possible to use workload partitioning
+to ensure the platform housekeeping processes don't run on the shared cpus.
+(This statement needs to be verified)
+
+* There is no risk that workload’s tasks run on dedicated OCP’s housekeeping reserved cpus,
+because only the shared cpus exposed via cgroups to the workload’s process.
+In other words, platform's housekeeping processes can run on shared cpus dedicated to workload’s light-weight tasks,
+but not the other way around.
+
+* mixed-cpus-node-plugin is NRI-based plugin.
+Bugs in the plugin or NRI framework might affect the containers at various lifecycle stages, e.g.,
+CreateContainer, UpdateContainer, etc.
+For mitigating such cases, it is possible to deactivate the node-plugin, and in severe cases it is possible to disable
+NRI completely at runtime-level.
+
+* From a performance perspective, the runtime communicates with the node-plugin via gRPC calls.
+Therefore, for every container creation, there's another gRPC call between the runtime and the node-plugin.
+We can't avoid this call even for containers that don't ask for shared CPUs, since runtime doesn't have any filtering
+mechanism, it sends all the containers through the NRI framework to node-plugin, and the filtering is done at the
```

Comment:
How is the lifecycle of the plugin managed? How does it affect the container runtime readiness advertised to the kubelet?

Reply:
@mrunalp 
> How is the lifecycle of the plugin managed? 
The plugin is a `DaemonSet` managed by NTO

> How does it affect the container runtime readiness advertised to the kubelet?
It does not. the plugin is eventually a pod and it won't interfere with the runtime readiness.
In case a pod that wants a shared cpus gets admitted, and the node-plugin pod is not ready, the pod keep pending until the node-plugin is up and running.
More about it here:
https://github.com/Tal-or/enhancements/blob/mixed-cpu-node-plugin/enhancements/node-tuning/mixed-cpu-node-plugin.md?plain=1#L132-L136

Reply:
is there a chicken/egg situation here?  what happens if the plugin is down, does it impact any guarantees that were provided to the workload that requested the alternative cpu?

Reply:
Is NRI an absolute requirement? Are inline hooks in CRI-O an alternative?

Reply:
@derekwaynecarr 
> what happens if the plugin is down, does it impact any guarantees that were provided to the workload that requested the alternative cpu?

If the plugin is down, the scheduler will try to place the workload on a node which has the plugin running and can fulfill the workload's resource requirement. 
In case the plugin is not running on any of the nodes and the workload is requesting an alternative cpu (i.e. share cpu) it keep pending and scheduler won't admit it.

This behavior is true for every workload which is asking for resources which cannot be fulfill by the cluster.

Reply:
@mrunalp 

>Is NRI an absolute requirement? Are inline hooks in CRI-O an alternative?

No, NRI is not a requirement and yes we can use inline hooks in CRI-O. 

We went with this approach for couple of reason:
1. NRI is runtime agnostic. Changes in CRI-O would make this feature a vendor lock-in feature, which we're trying to avoid.
2. We want to minimize regressions, so keeping CRI-O/Kubelet intact is a good way for achieving that.
3. NRI is modular, pluggable component and it provides a convenient way for injecting customize code during runtime.
Also please refer: https://github.com/openshift/enhancements/pull/1396#discussion_r1245690309